remote_attestation/rust/src/attestation.rs

```diff
@@ -15,238 +15,545 @@
+                .as_ref()
+                .context("Couldn't get attestation signer")?;
+            let tee_certificate = self
+                .behavior
+                .get_tee_certificate()
+                .as_ref()
+                .context("Couldn't get TEE certificate")?;
+
+            let attestation_info = create_attestation_info(&signer, &tee_certificate)
+                .context("Couldn't get attestation info")?;
+
+            let mut server_identity = ServerIdentity {
+                version: ATTESTATION_PROTOCOL_VERSION as i32,
+                ephemeral_public_key,
+                random: self.state.random.to_vec(),
+                transcript_signature: vec![],
+                signing_public_key: signer.public_key(),
+                attestation_info: Some(attestation_info),
+            };
+
+            // Update current transcript.
+            // Transcript doesn't include transcript signature from the server identity message.
+            self.transcript
+                .append(attestation_init)
+                .context("Couldn't append attestation init to transcript")?;
+            self.transcript
+                .append(&server_identity)
+                .context("Couldn't append server identity to transcript")?;
+
+            // Add transcript signature to the server identity message.
+            let transcript_signature = signer
+                .sign(&self.transcript.get_sha256())
+                .context("Couldn't create transcript signature")?;
+            server_identity.transcript_signature = transcript_signature;
+            server_identity
+        } else {
+            ServerIdentity {
+                version: ATTESTATION_PROTOCOL_VERSION as i32,
+                ephemeral_public_key,
+                random: self.state.random.to_vec(),
+                transcript_signature: vec![],
+                signing_public_key: vec![],
+                attestation_info: None,
+            }
+        };
+
+        let attestation_engine = self
+            .transition(key_negotiator)
+            .context("Couldn't transition from Initializing to Attesting state")?;
+        Ok((server_identity, attestation_engine))
+    }
+}
+
+impl AttestationEngine<Client, Attesting> {
+    /// Responds to `AttestationInit` message by creating a `ClientIdentity` message and derives
+    /// session keys for encrypting/decrypting messages from the server.
+    ///
+    /// `ClientIdentity` message contains an ephemeral public key for negotiating session keys.
+    /// If self attestation is enabled this message also provides necessary information to perform
+    /// remote attestation.
+    ///
+    /// Transitions [`AttestationEngine`] state from [`Attesting`] to [`Attested`] state.
+    pub fn process_server_identity(
+        mut self,
+        server_identity: &ServerIdentity,
+    ) -> anyhow::Result<(ClientIdentity, AttestationEngine<Client, Attested>)> {
+        if self.behavior.contains_peer_attestation() {
```

Comment:
Could this if-block be a `verify_server_identity` method in `AttestationBehavior` to make this method more readable?

Reply:
`AttestationBehavior` currently is server/client independant.
But the `verify_server_identity` will not be, since `ClientIdentity` and `ServerIdentity` messages are different (i.e. `ServerIdentity` has additional fields).